jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -0,0 +1,339 @@
+          if (macaroons !== null) {
+            headers['Macaroons'] = macaroons;
+          }
+          this.webhandler.sendGetRequest(
+            path, headers, null, null, false, null,
+            this._requestHandler.bind(
+              this, successCallback, failureCallback
+            )
+          );
+      },
+
+      /**
+       Handles the request response from the _makeRequest method, calling the
+       supplied failure callback if the response status was >= 400 or passing
+       the response object to the supplied success callback.
+
+       @method _requestHandler
+       @param {Function} successCallback Called when the api request completes
+       successfully.
+       @param {Function} failureCallback Called when the api request fails
+       with a response of >= 400.
+       @param {Object} response The XHR response object.
+       */
+      _requestHandler: function (successCallback, failureCallback, response) {
+        var target = response.target;
+        if (target.status >= 400) {
+          failureCallback(response);
+          return;
+        }
+        successCallback(response);
+      },
+
+      /**
+       Authenticate  by discharging the macaroon and
+       then set the cookie by calling the authCookiePath provided
+
+       @method authenticate
+       @param {Macaroon} The macaroon to be discharged
+       @param {String} The path where to send put request to set the cookie back
+       @param {Function} The request to be sent again in case of
+       successful authentication
+       @param {Function} The callback failure in case of wrong authentication
+       */
+      _authenticate: function (m, authCookiePath,
+                               requestFunction, failureCallback) {
+        try {
+          macaroon.discharge(
+            macaroon.import(m),
+            this._obtainThirdPartyDischarge.bind(this),
+            this._processDischarges.bind(
+              this,
+              authCookiePath,
+              requestFunction,
+              failureCallback
+            ),
+            failureCallback
+          );
+        } catch (ex) {
+          failureCallback(ex.message);
+        }
+      },
+
+      /**
+       Process the discharged macaroon and call the end point to be able to set
+       a cookie for the origin domain only when n auth cooie path i provided,
+       then call the original function.
+
+       @method _processDischarges
+       @param {String} The path where to send put request to set the cookie back
+       @param {Function} The request to be sent again in case of
+       successful authentication
+       @param {Function} The callback failure in case of wrong authentication
+       @param {[Macaroon]} The macaroons being discharged
+       */
+      _processDischarges: function (authCookiePath, requestFunction,
+                                    failureCallback, discharges) {
+        var jsonMacaroon = macaroon.export(discharges);
```

Comment:
I notice many of the other macaroon interactions use try/catch. If this fails will it also throw?